openshift/jenkins/Jenkinsfile

```diff
@@ -0,0 +1,323 @@
+                  apiSelector.startBuild("-w")
+                }
+              }
+            }
+          }
+        }
+      }
+      post {
+        success {
+          script {
+            env.API_BUILT = "true"
+          }
+        }
+      }
+    }
+    stage("Force FRONTEND Build?") {
+      agent any
+      steps {
+        script {
+          try {
+            timeout(time: 5, unit: "MINUTES") {
+              input(message: "Should we force the FRONTEND Build?", ok: "Yes, we should")
+            }
+            echo "[ci/cd]  The FRONTEND build was forced..."
+            env.FE_FORCE_BUILD = "true"
+          } catch (err) {
+            echo "[ci/cd]  No FRONTEND code changes. Skipping the FRONTEND build..."
+            env.FE_FORCE_BUILD = "false"
+          }
+        }
+      }
+    }
+    stage("Build Frontend") {
+      agent any
+      environment {
+        COMP_NAME = "app"
+        WAIT_TIMEOUT = 20 // minutes
+      }
+      when {
+        anyOf {
+          environment name: "FE_HAS_CHANGES", value: "true"
+          environment name: "FE_FORCE_BUILD", value: "true"
+        }
+      }
+      steps {
+        echo "[ci/cd]  Building FRONTEND artifacts..."
+        script {
+          openshift.withCluster() {
+            openshift.withProject(TOOLS_PROJECT) {
+              // Build the node application.
+              def build_config = "${APP_NAME}-${COMP_NAME}-base-${DEPLOY_TO}"
+              def feBase = openshift.selector("bc", build_config)
+              if (!feBase.exists()) {
+                throw new Exception("Could not find FRONTEND-base build with name: '${build_config}'")
+              } else {
+                timeout(WAIT_TIMEOUT) {
+                  feBase.startBuild("-w")
+                }
+              }
+
+              // Build the runtime image, if you are not using an off the shelf one.
+              def runtime_build_config = "${APP_NAME}-${COMP_NAME}-${DEPLOY_TO}"
+              def feRuntime = openshift.selector("bc", runtime_build_config)
+              if (!feRuntime.exists()) {
+                throw new Exception("Could not find FRONTEND-runtime build with name: '${runtime_build_config}'")
+              } else {
+                timeout(WAIT_TIMEOUT) {
+                  feRuntime.startBuild("-w")
+                }
+              }
+            }
+          }
+        }
+      }
+      post {
+        success {
+          script {
+            env.FE_BUILT = "true"
+          }
+        }
+      }
+    }
+    stage("Run Tests") {
+      agent any
+      when {
+        anyOf {
+          environment name: "API_BUILT", value: "true"
+          environment name: "FE_BUILT", value: "true"
+        }
+      }
+      steps {
+        echo "[ci/cd]  Tests should be run here"
+      }
+    }
+    stage("Deploy API to DEV") {
+      agent any
+      environment {
+        COMP_NAME = "api"
+      }
+      when { environment name: "API_BUILT", value: "true" }
```

Comment:
Wondering... When we deploy to a new environment for the first time we may not have changed the source, so API_BUILT and FE_BUILT may be false?

Reply:
This should handle the force build option.

Reply:
Correct, for that scenario a forced build is required. A forced build will update the API_BUILT and FE_BUILT flags appropriately